File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Auton.java
package org.firstinspires.ftc.teamcode;

import com.bylazar.configurables.annotations.Configurable;
import com.pedropathing.follower.Follower;
import com.pedropathing.geometry.BezierLine;
import com.pedropathing.geometry.Pose;
import com.pedropathing.paths.PathChain;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;

import org.firstinspires.ftc.teamcode.Helper.Hardware;
import org.firstinspires.ftc.teamcode.Helper.Shooter;
import org.firstinspires.ftc.teamcode.helpers.AprilTagConfig;
import org.firstinspires.ftc.teamcode.pedroPathing.Constants;

enum State {
    GOING_TO_PICKUP,
    PICKING_UP,
    GOING_TO_LAUNCH,
    DONE
}

@Configurable
public abstract class Auton extends LinearOpMode {
    // these poses are for the red side; blue will mirror them
    public static Pose start = new Pose(56, 8, 90);
    public static Pose pickupGPP = new Pose(40, 36, 180);
    public static Pose nextSetOffset = new Pose(0, 24);
    public static Pose lastBallOffset = new Pose(-16, 0);
    public static Pose launch = new Pose(56, 104, 35);

    @Override
    public void runOpMode() {
        Hardware.init(hardwareMap);
        Follower f = Constants.createFollower(hardwareMap);
        f.setStartingPose(start);
        AprilTagConfig atConf = new AprilTagConfig();
        Shooter shooter = new Shooter();
        while (!atConf.order.isPresent()) {
            atConf.Update();
            sleep(20);
        }
        Pose pickup = pickupGPP;
        switch (atConf.order.get()) {
            case PPG:
                pickup = pickup.plus(nextSetOffset);
            case PGP:
                pickup = pickup.plus(nextSetOffset);
        }
        telemetry.addData("Ball Order", atConf.order.get());
        telemetry.addLine();
        telemetry.update();
        waitForStart();
        PathChain toPickup = between(f, start, pickup);
        PathChain finishPickup = between(f, pickup, pickup.plus(lastBallOffset));
        PathChain toLaunch = between(f, pickup.plus(lastBallOffset), launch);
        State s = State.GOING_TO_PICKUP;
        f.followPath(toPickup);
        while (opModeIsActive()) {
            switch (s) {
                case GOING_TO_PICKUP:
                    if (!f.isBusy()) {
                        shooter.SetIntake(true);
                        f.followPath(finishPickup);
                        s = State.PICKING_UP;
                    }
                    break;
                case PICKING_UP:
                    if (!f.isBusy()) {
                        shooter.SetIntake(false);
                        f.followPath(toLaunch, true);
                        s = State.GOING_TO_LAUNCH;
                    }
                    break;
                case GOING_TO_LAUNCH:
                    if (!f.isBusy()) {
                        shooter.SetOuttake(true);
                        s = State.DONE;
                    }
                    break;
            }
        }
    }

    // red should just return p, blue should mirror p
    abstract Pose transform(Pose p);

    PathChain between(Follower f, Pose begin, Pose end) {
        begin = transform(begin);
        end = transform(end);
        return f.pathBuilder().addPath(new BezierLine(begin, end)).setLinearHeadingInterpolation(begin.getHeading(), end.getHeading()).build();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Auton.kt
package org.firstinspires.ftc.teamcode

import com.pedropathing.follower.Follower
import com.pedropathing.geometry.BezierLine
import com.pedropathing.geometry.Pose
import com.pedropathing.paths.PathChain
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import org.firstinspires.ftc.teamcode.helper.AprilTagConfig
import org.firstinspires.ftc.teamcode.helper.BallOrder
import org.firstinspires.ftc.teamcode.helper.Hardware
import org.firstinspires.ftc.teamcode.helper.Shooter
import org.firstinspires.ftc.teamcode.pedroPathing.Constants

internal enum class State {
    GOING_TO_PICKUP,
    PICKING_UP,
    RETURN_TO_PICKUP,
    GOING_TO_LAUNCH,
    DONE
}

abstract class Auton : LinearOpMode() {
    abstract fun getWaypoints(): Waypoints
    override fun runOpMode() {
        val w = getWaypoints()
        Hardware.init(hardwareMap)
        val f = Constants.createFollower(hardwareMap)
        f.setStartingPose(w.start)
        val atConf = AprilTagConfig()
        val shooter = Shooter()
        while (atConf.order == null) {
            atConf.Update()
            sleep(20)
        }
        val pickup = w.pickupGPP.plus(
            when (atConf.order!!) {
                BallOrder.PPG -> w.nextSetOffset * 2.0
                BallOrder.PGP -> w.nextSetOffset
                BallOrder.GPP -> Pose(0.0, 0.0, 0.0)
            }
        )
        val pickupEnd = pickup.plus(w.lastBallOffset)
        telemetry.addData("Ball Order", atConf.order)
        telemetry.addLine()
        telemetry.update()
        waitForStart()
        val toPickup = between(f, w.start, pickup)
        val finishPickup = between(f, pickup, pickupEnd)
        val returnToPickup = between(f, pickupEnd, pickup)
        val toLaunch = between(f, pickupEnd, w.launch)
        var s = State.GOING_TO_PICKUP
        f.followPath(toPickup)
        while (opModeIsActive()) {
            when (s) {
                State.GOING_TO_PICKUP -> if (!f.isBusy) {
                    shooter.intake = true
                    f.followPath(finishPickup)
                    s = State.PICKING_UP
                }

                State.PICKING_UP -> if (!f.isBusy) {
                    shooter.intake = false
                    f.followPath(returnToPickup)
                    s = State.RETURN_TO_PICKUP
                }

                State.RETURN_TO_PICKUP -> if (!f.isBusy) {
                    f.followPath(toLaunch, true)
                    s = State.GOING_TO_LAUNCH
                }

                State.GOING_TO_LAUNCH -> if (!f.isBusy) {
                    shooter.outtake = true
                    s = State.DONE
                }

                State.DONE -> {}
            }
        }
    }

    fun between(f: Follower, begin: Pose, end: Pose): PathChain? {
        return f.pathBuilder().addPath(BezierLine(begin, end))
            .setLinearHeadingInterpolation(begin.heading, end.heading).build()
    }

    data class Waypoints(
        @JvmField
        val start: Pose,
        @JvmField
        val pickupGPP: Pose,
        @JvmField
        val nextSetOffset: Pose,
        @JvmField
        val lastBallOffset: Pose,
        @JvmField
        val launch: Pose
    )
}